### PR TITLE
ImageBufferIOSurfaceBackend should flush only when GraphicsContextCG has drawn

### PR DESCRIPTION
**Pull request by kkinnunen-apple**
#### 9e8259e401bec8f175f796fac39c34888bcbc5cb
<pre>
ImageBufferIOSurfaceBackend should flush only when GraphicsContextCG has drawn
<a href="https://bugs.webkit.org/show_bug.cgi?id=253309">https://bugs.webkit.org/show_bug.cgi?id=253309</a>
rdar://106192612

Reviewed by Simon Fraser.

Track draws in GraphicsContextCG. Currently it overestimates when the
draw happens.

Use this information in ImageBufferIOSurfaceBackend. The backend
needs to know if the surface has been drawn to before doing an external
operation. If draw to has happened, the drawing context needs to be
flushed before external operation can happen. Conversely, the flush
can be skipped if we know that there was no drawing.

Consider case:
i1.draw(something);   // Initial image
i1.getPixelBuffer();  // Flush
i1.putPixelBuffer();  // No flush needed, no draws since flush. &lt;-- example of skipped flush.
i2.draw(i1);          // Creates an image from i1 (flushes currently).

i1.getPixelBuffer();  // No flush &lt;-- example of skipped flush.
i1.putPixelBuffer();  // Invalidates, flush needed due to invalidate, no draws.
i2.draw(i1);          // Creates an image from i1 (flushes currently).

* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::platformContext const):
(WebCore::GraphicsContextCG::contextForDraw):
(WebCore::GraphicsContextCG::contextForState const):
(WebCore::GraphicsContextCG::save):
(WebCore::GraphicsContextCG::restore):
(WebCore::GraphicsContextCG::consumeHasDrawn):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::context):
(WebCore::ImageBufferIOSurfaceBackend::flushContext):
(WebCore::ImageBufferIOSurfaceBackend::flushContextIfNeeded):
(WebCore::ImageBufferIOSurfaceBackend::invalidateCachedNativeImage):
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImage):
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImageForDrawing):
(WebCore::ImageBufferIOSurfaceBackend::prepareForExternalRead):
(WebCore::ImageBufferIOSurfaceBackend::prepareForExternalWrite):
(WebCore::ImageBufferIOSurfaceBackend::ensureSurfaceContext):
(WebCore::ImageBufferIOSurfaceBackend::createImage):
(WebCore::ImageBufferIOSurfaceBackend::createImageReference):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Tools/TestWebKitAPI/Tests/WebCore/cg/GraphicsContextCGTests.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/262010@main">https://commits.webkit.org/262010@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/6f349f1bf29aacdad394462ddf3f5800693d11d5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/95 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/97 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/77 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/90 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/89 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/72 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/79 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/73 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/86 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/82 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/72 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/68 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/62 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/86 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp [hunk 1]
```diff
@@ -1555,6 +1567,15 @@ bool GraphicsContextCG::canUseShadowBlur() const
     return (renderingMode() == RenderingMode::Unaccelerated) && hasBlurredShadow() && !m_state.shadowsIgnoreTransforms();
 }
 
+bool GraphicsContextCG::commitDraws()
```
Comment (hortont424) [inline Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp]: Is this maybe `didFlushDrawing`? Since it doesn't actually commit anything.

Comment (kkinnunen-apple) [reply Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp]: No, this is completely inverse of "did flush drawing".  Method named "didFlushDrawing" would be called as signal, "I did flush the drawing, do what you need to do". This is totally inverse of what is intended here.

In case my English is poor and you meant "didFlushDrawing" in the meaning of "drawingWasFlushed", it is still the same: the intention and the code as it reads is totally inverse this.

This is the signal for the caller to decide if they need to flush the drawing.

TLDR; the commitDraw() commits the state "did we draw" (as well as potential buffered draws).

The caller here is interested in chunks of time which they decide based on some arbitrary criteria.

So the sequence they query is like:
* commitDraws()
* commitDraws()
* commitDraws()
( This point is to highlight that that is the only interface this part of the caller uses -- it divides the time with these calls and cannot access anything else).

Those commits say to the drawing context :
- please commit all the draws that happened since last time I called
- please tell me if there was a draw that was committed

Whether or not the drawing context *does* any work *today* for this trigger is encapsulated inside the implementation, it is not concern of the interface.

If the implementation was such that it would optimise the draws that negate some previous draws, it might build a buffer of draws. This is the place such an implementation would need to commit the buffer instead of build up more for that purpose. 

Continuing: all buffers have a max length.  So in above scenario, of course the work would be sent further either when the buffer would fill up or when the commit needs it. So obviously in above hypothetical implementation the work would be sent forward not only at commit, but during draws.

The flag represents the state of above mentioned buffer, but with the distinction that the buffer is actually zero length -- e.g. it does not have capacity to defer any of the mentioned draws, but it still can track whether there was a draw or not. So thinking in other way, it commits the deferred draw buffer where the buffer length is zero..

File: Source/WebCore/platform/graphics/cg/GraphicsContextCG.h [hunk 2]
```diff
@@ -127,6 +129,15 @@ class WEBCORE_EXPORT GraphicsContextCG : public GraphicsContext {
 
     virtual FloatRect roundToDevicePixels(const FloatRect&, RoundingMode = RoundAllSides) const;
 
+    // Returns the platform context for draws.
+    CGContextRef contextForDraw();
```
Comment (smfr) [inline Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:135]: Maybe `contextForDrawing()`

It's going to be easy to make a mistake in future and get the context the wrong way. How about passing a flag to `platformContext()`, so we have to force people to choose a flag at all call sites?

File: Source/WebCore/platform/graphics/cg/GraphicsContextCG.h [hunk 3]
```diff
@@ -140,6 +151,8 @@ class WEBCORE_EXPORT GraphicsContextCG : public GraphicsContext {
     const RenderingMode m_renderingMode : 1; // NOLINT
     const bool m_isLayerCGContext : 1;
     mutable bool m_userToDeviceTransformKnownToBeIdentity : 1 { false };
+    // Flag for pending draws. Start with true because we do not know what commands have been scheduled to the context.
+    bool m_hasDrawn : 1 { true };
```
Comment (smfr) [inline Source/WebCore/platform/graphics/cg/GraphicsContextCG.h]: Would prefer a name that indicates when this bit gets cleared, like `m_hasDrawnSinceLastCommit` or whatever.

File: Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp (hunk 1 above)
Comment (smfr) [reply Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp]: I'm also confused by the naming here. It sounds like `commitDraws()` would actually do work, but it just resets the flag. This is really just `std::exchange(m_hasDrawn, false)`. Maybe the call sites should just use that?

File: Source/WebCore/platform/graphics/cg/GraphicsContextCG.h (hunk 2 above)
Comment (kkinnunen-apple) [reply Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:135]: > Maybe `contextForDrawing()`

Done.

> It's going to be easy to make a mistake in future and get the context the wrong way. How about passing a flag to `platformContext()`, so we have to force people to choose a flag at all call sites?

I think with current code it's actually very *hard* to make a mistake. `platformContext()` will use the the correct conservative function, `contextForDrawing()`. So if it goes wrong, it's only a very slight perf problem. Theoretical one, as the sequence is such that the context accessors are relevant for theming code, and that is very far away from the place where the draw bookkeeping would optimise anything.

Currently there are no clients outside GraphicsContextGC that can in practice pessimize the flushes, and probably won't be.  Future commits will remove platformContext() usage from GraphicsContextGC. Future future commits, hopefully, will remove platformContext() usage from the rest of CG codepaths (well, probably a stretch.. but in spirit).

The function `GraphicsContext::platformContext()` is kind of a bug in and of itself. I would want to work on removing it instead of modifying, as it is error prone and needs redundancy in form of `hasPlatformContext()`.

Contrasting:

```
if (!c.hasPlatformContext())
   return;
// Hopefully this is CG-based??
c.platformContext(ForDrawing);
c.platformContext(ForState);
```
vs.
```
if (!is<GraphicsContextCG>(c))
   return;
auto& context = downcast<GraphicsContextCG>(c);
context.contextForDrawing();
context.contextForState();
```
I think the latter is the sane approach and the direction we should put the effort in.

In other words:
 - `platformContext()` function is a bug, since  not all `GraphicsContexts` have `platformContext()` or there's no single universal `platformContext()` for a project. Hence it does not make sense to have such a function. Leads to situation that we have now where `GraphicsContext` clients have all sorts of (laborsome?) nullability checks and string prints but still also fail to correctly do the checking, rendering the previous nullability checks confusing.
 - I already added the correct way to access the context: two *different* functions for different purposes. It is IMO better than one function that turns into two distinct functions based on the two-value argument being passed. If the currently proposed approach wasn't better, it would get a review comment of form "would you merge your `contextForState` and `contextForDrawing` into one function with a parameter `contextFor(Case)`. I don't feel like that would be the correct pattern, hence I started with the two functions. However, also in this the correct solution is not to overload the old platformContext() that is opaque function across all the `GraphicsContexts` -- there are `GraphicsContexts` that do not need the `Case` parameter. Rather that would be a function in GraphicsContextCG if I'm not mistaken?

File: Source/WebCore/platform/graphics/cg/GraphicsContextCG.h (hunk 3 above)
Comment (kkinnunen-apple) [reply Source/WebCore/platform/graphics/cg/GraphicsContextCG.h]: done.

File: Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp [hunk 4]
```diff
@@ -1571,6 +1583,15 @@ bool GraphicsContextCG::canUseShadowBlur() const
...
+{
+    if (m_hasDrawnSinceLastCommit) {
+        m_hasDrawnSinceLastCommit = false;
```
Comment (mattwoodrow) [inline Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp]: It seems slightly sad that a caller could still have access to the `CGContextRef` obtained from `platformContext` when this is called, and then add new draw calls later that wouldn't be tracked.

It would be nice if the public API returned some sort of 'borrowed' object, that we could assert has been returned/destroyed before this was called.

I guess that still doesn't help too much, since there are still other ways for external users to have the `CGContextRef` (like from the IOSurface), and thus we have to still manually determine external reasons to flush (or that flushing has been done for us).

Without large changes to make this a much more closed abstraction around the `CGContext`, then this seems like the best way can do as a 'have there been any draw calls visible to this object, since the last time this was called'.

Not sure what the resolution is for the naming discussion, but the approach seems reasonable to me.

Comment (kkinnunen-apple) [reply Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp]: >It would be nice if the public API returned some sort of 'borrowed' object, that we could assert has been returned/destroyed before this was called.

This is explored in #11354 [11354](https://github.com/WebKit/WebKit/pull/11354)

> I guess that still doesn't help too much, since there are still other ways for external users to have the CGContextRef (like from the IOSurface), and thus we have to still manually determine external reasons to flush (or that flushing has been done for us).

I think this is a bug in IOSurface -- there's no use-case for such a construct as to store the stale or random CGContext from the IOSurface. IOSurface::ensurePlatformContext() should be renamed as IOSurface::createPlatformContext().

File: Source/WebCore/platform/graphics/cg/GraphicsContextCG.h [hunk 5]
```diff
@@ -129,6 +131,15 @@ class WEBCORE_EXPORT GraphicsContextCG : public GraphicsContext {
 
     virtual FloatRect roundToDevicePixels(const FloatRect&, RoundingMode = RoundAllSides) const;
 
+    // Returns the platform context for draws.
+    CGContextRef contextForDrawing();
+
+    // Returns the platform context for purposes of context state change, not draws.
+    CGContextRef contextForState() const;
```
Comment (smfr) [inline Source/WebCore/platform/graphics/cg/GraphicsContextCG.h]: Can this be private? That cleans up the interface nicely, and maybe we don't even need `contextForDrawing()` as public, since all public callers will use the context for drawing?

File: Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp (hunk 1 above)
Comment (smfr) [reply Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp]: Still not very happy with the naming here. Maybe the more literal `takeHasDrawnSinceLastCommit()` or a pair: `hasDrawnSinceLastCommit()` and `clearHasDrawnSinceLastCommit()` or something with "needs flush" or "will flush" in the name. It needs to indicate that it comes before some code that either implicitly or explicitly triggers a CGContextFlush

File: Source/WebCore/platform/graphics/cg/GraphicsContextCG.h (hunk 2 above)
Comment (smfr) [reply Source/WebCore/platform/graphics/cg/GraphicsContextCG.h:135]: You still have `contextForDraw()` in this PR.

File: Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp [hunk 6]
```diff
@@ -274,23 +275,54 @@ void ImageBufferIOSurfaceBackend::prepareForExternalRead()
...
+        needFlush = true;
     }
 
     // Ensure that there are no pending draws to this surface. This is ensured by flushing the context
     // through which the draws may have come.
     // Ensure that there are no pending draws from this surface. This is ensured by drawing the invalidation marker before
     // flushing the the context. The invalidation marker forces the draws from this surface to complete before
     // the invalidation marker completes.
-    flushContext();
+    if (flushContextIfNeeded())
```
Comment (smfr) [inline Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp]: It's pretty confusing that this function calls both `flushContextIfNeeded()` and `CGContextFlush()`.

**Review by smfr — APPROVED** [review]

File: Source/WebCore/platform/graphics/cg/GraphicsContextCG.h (hunk 5 above)
Comment (kkinnunen-apple) [reply Source/WebCore/platform/graphics/cg/GraphicsContextCG.h]: moved contextForState() private.
Preserved contextForDrawing() as that should be used by components outside GraphicsContextCG that do drawing on the context.

File: Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp (hunk 6 above)
Comment (kkinnunen-apple) [reply Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp]: Renamed `flushContextIfNeeded` to `flushContextDraws`. Taking the liberty to move forward to submitting, I can rename this if it is not a good name.
